### PR TITLE
Added a contributing file

### PR DESCRIPTION
**Pull request by blaisb**
This PR adds a CONTRIBUTING.md file to give some guidelines on contributions. I added a lot of you as reviewers. Please offer as much feedback as possible to make this file as neat and tidy as you can. You can use the github review interface.

### CODE REVIEW

**Review by jeannejoachim — COMMENTED** [review]
Very clear! The only misleading point that I find is in the good practice : "Classes should be implemented in their `.cc`". Isn't it only for non-template classes? Lethe uses a lot of template classes, whose implementation is also directly in the `.h`, with no `.cc`. If I understood it right, please maybe clarify this point :)

**Comment by blaisb** [comment]
> Very clear! The only misleading point that I find is in the good practice : "Classes should be implemented in their `.cc`". Isn't it only for non-template classes? Lethe uses a lot of template classes, whose implementation is also directly in the `.h`, with no `.cc`. If I understood it right, please maybe clarify this point :)

Good point. Actually the classes implemented in their '.h' right now are due to laziness. Eventually most of them should be switched to '.cc' except for inlined functions and trivial getters :)!

**Comment by acdaigneault** [comment]
As Joanne said, it's very clear! 
I have a suggestion, not about contribution but about helping to use Lethe for the first time. 
It should be nice to have a file explaining the solvers and the possible options. I know there's the wiki about the parameters, but it's not very explicit (to me). I'm talking about a file with the name of solvers and a short explanation about their strengths/applications with a bulleted list of principal options & features with the related examples. I think it will helps to have a big picture of the software. If it's not for new users, I should be useful for new students.
I don't know if my explainations are clear, let me know. :)

**Comment by jeannejoachim** [comment]
> > Very clear! The only misleading point that I find is in the good practice : "Classes should be implemented in their `.cc`". Isn't it only for non-template classes? Lethe uses a lot of template classes, whose implementation is also directly in the `.h`, with no `.cc`. If I understood it right, please maybe clarify this point :)
> 
> Good point. Actually the classes implemented in their '.h' right now are due to laziness. Eventually most of them should be switched to '.cc' except for inlined functions and trivial getters :)!

Is it really feasible for template classes? In the C++ course I took, it is well precised that for a template class, everything must be in the `.h` : https://openclassrooms.com/fr/courses/1894236-programmez-avec-le-langage-c/1903999-creez-des-templates#/id/r-1913608

**Comment by jeannejoachim** [comment]
sorry this was the link for template functions, here is the link for template classes : https://openclassrooms.com/fr/courses/1894236-programmez-avec-le-langage-c/1903999-creez-des-templates#/id/r-1913691

**Comment by acdaigneault** [comment]
> > > Very clear! The only misleading point that I find is in the good practice : "Classes should be implemented in their `.cc`". Isn't it only for non-template classes? Lethe uses a lot of template classes, whose implementation is also directly in the `.h`, with no `.cc`. If I understood it right, please maybe clarify this point :)
> > 
> > 
> > Good point. Actually the classes implemented in their '.h' right now are due to laziness. Eventually most of them should be switched to '.cc' except for inlined functions and trivial getters :)!
> 
> Is it really feasible for template classes? In the C++ course I took, it is well precised that for a template class, everything must be in the `.h` : https://openclassrooms.com/fr/courses/1894236-programmez-avec-le-langage-c/1903999-creez-des-templates#/id/r-1913608

I asked Bruno about that this week. You have to fill the template for the class in the .cc file (like template class TestClass<2, TrininosWrappers::MPI::Vector, IndexSet>) for all the possibilities. For example, look at the bottom of the NavierStokesBase class. :)

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Lethe
+
+Lethe is an open source and we try to follow the deal.II mentality of having it
+developed an open accessible environment. Consequently, all scientific
```
Comment (oguevremontmec8211) [inline CONTRIBUTING.md:4]: Add "in"

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+development made within Lethe are accessible to everyone even as they are
```
Comment (oguevremontmec8211) [inline CONTRIBUTING.md:5]: Maybe "developments" since the following verbs are plural.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+currently being written. Anybody that wishes to use and contribute to a feature
+is welcomed to do so.
```
Comment (oguevremontmec8211) [inline CONTRIBUTING.md:7]: I'd change to "welcome".

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+This document aims at establishing from guidelines for contributions within
```
Comment (oguevremontmec8211) [inline CONTRIBUTING.md:9]: I'd remove "from"

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+used to open pull requests
+
+# What should be the content of a pull request (PR)?
+
+A pull request should contain the following elements:
+
+- A brief title (less than 60 characters) that describes the goal of the pull
+requests
+- A detailed description of the content added by the pull request:
+i) If the PR adds a new feature. The feature should be documented, tests (with
```
Comment (oguevremontmec8211) [inline CONTRIBUTING.md:28]: The sentence should continue, so I'd change to "," instead of "."

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  a unit tests and/or an application test) and if the features adds or alters
```
Comment (oguevremontmec8211) [inline CONTRIBUTING.md:29]: I'd remove the "s" from feature, or change the uses of the word in the rest of the paragraph.

**Review by oguevremontmec8211 — COMMENTED** [review]
The content of the file makes sense and is complete in my opinion. The comments I have concern grammar only, but the text is clear as it is.

**Review by oguevremontmec8211 — APPROVED** [review]
Here is the approval, I don't think I selected properly before.

**Comment by Luckabarbeau** [comment]
Very clear ! I would also add a small line talk link to the wiki and also put the information content in this text file in the wiki.

**Review by shahabgol — APPROVED** [review]
Comprehensive and clear.

File: CONTRIBUTING.md (hunk 1 above)
Comment (oguevremont) [inline CONTRIBUTING.md:4]: Add "in"

File: CONTRIBUTING.md (hunk 2 above)
Comment (oguevremont) [inline CONTRIBUTING.md:5]: Maybe "developments" since the following verbs are plural.

File: CONTRIBUTING.md (hunk 3 above)
Comment (oguevremont) [inline CONTRIBUTING.md:7]: I'd change to "welcome".

File: CONTRIBUTING.md (hunk 4 above)
Comment (oguevremont) [inline CONTRIBUTING.md:9]: I'd remove "from"

File: CONTRIBUTING.md (hunk 5 above)
Comment (oguevremont) [inline CONTRIBUTING.md:28]: The sentence should continue, so I'd change to "," instead of "."

File: CONTRIBUTING.md (hunk 6 above)
Comment (oguevremont) [inline CONTRIBUTING.md:29]: I'd remove the "s" from feature, or change the uses of the word in the rest of the paragraph.

**Review by oguevremont — APPROVED** [review]
The content of the file makes sense and is complete in my opinion. The comments I have concern grammar only, but the text is clear as it is.

**Pull request merged**